x/bank/src/keeper.rs

```diff
@@ -372,6 +557,216 @@ impl<SK: StoreKey, PSK: ParamsSubspaceKey, AK: AuthKeeper<SK, M>, M: Module>
+                // TODO: serialization of vector of coins
+                EventAttribute::new(
+                    "amount".into(),
+                    serde_json::to_vec(&amount)
+                        .unwrap_or(amount.inner()[0].encode_vec().expect(IBC_ENCODE_UNWRAP))
+                        .into(),
+                    true,
+                ),
+            ],
+        ));
+
+        Ok(self.add_coins(ctx, &module_acc_addr, balances)?)
+    }
+
+    /// undelegate_coins performs undelegation by crediting amt coins to an account with
+    /// address addr. For vesting accounts, undelegation amounts are tracked for both
+    /// vesting and vested coins. The coins are then transferred from a ModuleAccount
+    /// address to the delegator address. If any of the undelegation amounts are
+    /// negative, an error is returned.
+    fn undelegate_coins<DB: Database, CTX: TransactionalContext<DB, SK>>(
+        &self,
+        ctx: &mut CTX,
+        module_acc_addr: AccAddress,
+        delegator_addr: AccAddress,
+        amount: SendCoins,
+    ) -> Result<(), AppError> {
+        if self
+            .auth_keeper
+            .get_account(ctx, &module_acc_addr)?
+            .is_none()
+        {
+            return Err(AppError::AccountNotFound);
+        };
+
+        self.sub_unlocked_coins(ctx, &module_acc_addr, &amount)?;
+        self.track_undelegation(ctx, &delegator_addr, &amount)?;
+        Ok(self.add_coins(ctx, &delegator_addr, amount.into_inner())?)
+    }
+
+    /// sub_unlocked_coins removes the unlocked amt coins of the given account. An error is
+    /// returned if the resulting balance is negative or the initial amount is invalid.
+    /// A coin_spent event is emitted after.
+    fn sub_unlocked_coins<DB: Database, CTX: TransactionalContext<DB, SK>>(
+        &self,
+        ctx: &mut CTX,
+        addr: &AccAddress,
+        amount: &SendCoins,
+    ) -> Result<(), AppError> {
+        let locked_coins = self.locked_coins(ctx, addr)?;
+
+        let amount_of = |coins: &Vec<Coin>, denom: &Denom| -> Uint256 {
+            let coins = coins.iter().find(|c| c.denom == *denom);
+            coins.map(|c| c.amount).unwrap_or(Uint256::zero())
+        };
+
+        for coin in amount.inner() {
+            if let Some(mut balance) = self.balance(ctx, addr, &coin.denom)? {
+                let locked_amount = amount_of(&locked_coins, &coin.denom);
+                let spendable = balance.amount - locked_amount;
+
+                if spendable.checked_sub(coin.amount).is_err() {
+                    return Err(AppError::Coins(format!(
+                        "{} is smaller than {}",
+                        spendable, coin.amount
+                    )));
+                }
+
+                balance.amount -= coin.amount;
+                self.set_balance(ctx, addr, balance)?;
+            } else {
+                return Err(AppError::Coins(format!(
+                    "Cannot get balance of account {}",
```

Comment:
It's not that we can't get the balance of account... It's that there is no entry for that coin which means they don't have sufficient funds.